internal/runners/builds/download.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/internal/runbits/rationalize"
+	"github.com/ActiveState/cli/internal/runbits/runtime"
+	"github.com/ActiveState/cli/internal/runbits/runtime/progress"
+	auth "github.com/ActiveState/cli/pkg/platform/authentication"
+	"github.com/ActiveState/cli/pkg/platform/model"
+	"github.com/ActiveState/cli/pkg/platform/runtime/artifact"
+	"github.com/ActiveState/cli/pkg/platform/runtime/buildplan"
+	rtProgress "github.com/ActiveState/cli/pkg/platform/runtime/setup/events/progress"
+	"github.com/ActiveState/cli/pkg/platform/runtime/store"
+	"github.com/ActiveState/cli/pkg/platform/runtime/target"
+	"github.com/ActiveState/cli/pkg/project"
+	"github.com/vbauerster/mpb/v7"
+	"github.com/vbauerster/mpb/v7/decor"
+)
+
+type DownloadParams struct {
+	BuildID   string
+	OutputDir string
+}
+
+type Download struct {
+	out       output.Outputer
+	project   *project.Project
+	analytics analytics.Dispatcher
+	svcModel  *model.SvcModel
+	auth      *auth.Auth
+	config    *config.Instance
+}
+
+func NewDownload(prime primeable) *Download {
+	return &Download{
+		out:       prime.Output(),
+		project:   prime.Project(),
+		analytics: prime.Analytics(),
+		svcModel:  prime.SvcModel(),
+		auth:      prime.Auth(),
+		config:    prime.Config(),
+	}
+}
+
+func (d *Download) Run(params *DownloadParams) (rerr error) {
+	defer rationalizeError(&rerr)
+
+	if d.project == nil {
+		return rationalize.ErrNoProject
+	}
+
+	_, err := runtime.NewFromProject(d.project, target.TriggerBuilds, d.analytics, d.svcModel, d.out, d.auth, d.config)
+	if err != nil {
+		return locale.WrapInputError(err, "err_refresh_runtime_new", "Could not update runtime for this project.")
+	}
+
+	runtimeStore := store.New(target.NewProjectTarget(d.project, nil, target.TriggerBuilds).Dir())
+	bp, err := runtimeStore.BuildPlan()
+	if err != nil {
+		return errs.Wrap(err, "Could not get build plan")
+	}
+
+	terminalArtfMap, err := buildplan.NewMapFromBuildPlan(bp, false, false, nil)
+	if err != nil {
+		return errs.Wrap(err, "Could not get build plan map")
+	}
+
+	var artifact *artifact.Artifact
+	for _, artifacts := range terminalArtfMap {
+		for _, a := range artifacts {
+			if strings.HasPrefix(strings.ToLower(string(a.ArtifactID)), strings.ToLower(params.BuildID)) {
+				artifact = &a
+				break
+			}
+		}
+	}
+
+	if artifact == nil {
+		return locale.NewInputError("err_build_id_not_found", "Could not find artifact with ID {{.V0}}", params.BuildID)
+	}
+
+	targetDir := params.OutputDir
+	if targetDir == "" {
+		targetDir, err = os.Getwd()
+		if err != nil {
+			return errs.Wrap(err, "Could not get current working directory")
+		}
+	}
+
+	if err := d.downloadArtifact(artifact, targetDir); err != nil {
+		return errs.Wrap(err, "Could not download artifact %s", artifact.ArtifactID.String())
+	}
+
+	return nil
+}
+
+func (d *Download) downloadArtifact(artifact *artifact.Artifact, targetDir string) (rerr error) {
```

Comment:
I implemented this only to find that the spinner does not produce any output on non-interactive: https://github.com/ActiveState/cli/blob/master/internal/output/progress.go#L36. Confirmed with Mitchell on slack that we only do progress on plain output as we want to ensure any output is valid JSON.

Reply:
Non-interactive is not the same as structured. Plain output can be either interactive or non-interactive. For example CI would be non-interactive, but would still need some form of progress indication otherwise most CI systems would consider the task "hung" after a certain amount of time.

Reply:
That makes sense. Do we want to update the spinner in this PR? I feel like it needs its own story as it could have further reaching implications.

Reply:
No I wouldn't want to update it for this PR, but if you only invoke it when in non-interactive mode then it should still give us the desired behavior.